### PR TITLE
Rust gRPC + Protobuf Bindings

### PR DESCRIPTION
**Pull request by alexjpwalker**
## What is the goal of this PR?

We've added a gRPC + Protobuf compiler for Rust, allowing us to generate Rust bindings for our Protobuf definitions using the `grpc` crate.

## What are the changes implemented in this PR?

We experimented with a couple of different approaches.

First, I tried implementing the compiler using the `tonic` crate. This is the most popular gRPC crate for Rust, and has by far the most recent GitHub activity. However, a limitation of the compiler library I was using appeared to make it impossible to compile multiple `.proto` files at once, which became a blocker. It's possible that that functionality _is_ provided, but to save time, I moved onto the next approach:

Second, I tried implementing the compiler using the `grpc` crate. This is the second most popular gRPC crate for Rust, and has 1.5k GitHub stars (vs `tonic`'s 5k). This one was completed without difficulties, and I've checked it out into the repo in this PR as reference, but we do not intend to use it going forward.

Finally, I found a `rust_grpc_library` rule in `rules_rust`. This is our preferential solution as it delegates all the work of writing the compilation script to an externally-maintained codebase. It doesn't require us to write a single line of Rust code manually, it just works out of the box.

The only limitation is, it is built for the `grpc` crate, our second-favourite crate for the job. We are currently researching whether we can achieve an elegant and working solution using the `tonic` crate.

However, this PR has been open for a while now, and as we have other projects beginning to depend on it - e.g. `factory` - it makes sense to merge these changes upstream and iterate further on it from there.

### CODE REVIEW

File: grpc/rust/BUILD [hunk 1]
```diff
@@ -0,0 +1,39 @@
+load("@rules_rust//proto:proto.bzl", "rust_grpc_library")
+load("@rules_rust//proto:toolchain.bzl", "rust_proto_toolchain")
+
+rust_proto_toolchain(
```
Comment (alexjpwalker) [inline grpc/rust/BUILD:4]: Source: https://bazelbuild.github.io/rules_rust/rust_proto.html#customizing-dependencies
> These rules depends on the [protobuf](https://crates.io/crates/protobuf) and the [grpc](https://crates.io/crates/grpc) crates in addition to the [protobuf compiler](https://github.com/google/protobuf). To obtain these crates, rust_proto_repositories imports the given crates using BUILD files generated with [cargo raze](https://github.com/google/cargo-raze).
>
> If you want to either change the protobuf and gRPC rust compilers, or to simply use [cargo raze](https://github.com/google/cargo-raze) in a more complex scenario (with more dependencies), you must redefine those dependencies.
>
> To do this, once you’ve imported the needed dependencies (see our [Cargo.toml](https://bazelbuild.github.io/rules_rust/raze/Cargo.toml) file to see the default dependencies), you need to create your own toolchain. To do so you can create a BUILD file with your toolchain definition

Comment (alexjpwalker) [reply grpc/rust/BUILD:4]: TO REVISIT: Where should this toolchain be kept? Architecturally, it should probably live in `@vaticle_dependencies`.

File: WORKSPACE [hunk 2]
```diff
@@ -46,6 +46,17 @@ kt_register_toolchains()
 load("@vaticle_dependencies//builder/python:deps.bzl", python_deps = "deps")
 python_deps()
 
+# Load //builder/rust
```
Comment (alexjpwalker) [inline WORKSPACE:49]: We add the Rust toolchain and fetch crate dependencies of the gRPC + Protobuf Rust compiler.

File: grpc/rust/main.rs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+extern crate protoc_rust_grpc;
```
Comment (alexjpwalker) [inline grpc/rust/main.rs:1]: This was my second attempt at getting a working gRPC + Protobuf Rust compiler binary.

It works, but, ideally we would like to not write our own `main.rs`, but rather delegate it to an established Bazel Rust rules library - ie: `rules_rust` or other.

File: dependencies/vaticle/repositories.bzl [hunk 4]
```diff
@@ -21,5 +21,5 @@ def vaticle_dependencies():
     git_repository(
         name = "vaticle_dependencies",
         remote = "https://github.com/vaticle/dependencies",
-        commit = "a8b3b714a5e0562d41f4c05ca8f266d48b7d0fd3", # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_dependencies
+        commit = "e3ba4970a3bb7db5e4725c0daa2a2a154ba9b3b8", # sync-marker: do not remove this comment, this is used for sync-dependencies by @vaticle_dependencies
```
Comment (alexjpwalker) [inline dependencies/vaticle/repositories.bzl:24]: New crate dependencies used by the gRPC + Protobuf compiler for Rust.

File: grpc/rust/BUILD [hunk 5]
```diff
@@ -0,0 +1,64 @@
+load("@rules_rust//proto:proto.bzl", "rust_grpc_library")
+load("@rules_rust//proto:toolchain.bzl", "rust_proto_toolchain")
+load("@rules_rust//rust:defs.bzl", "rust_binary")
+
+# TODO: once we clarify how we should compile the gRPC + Protobuf library, delete this target
+rust_binary(
```
Comment (alexjpwalker) [inline grpc/rust/BUILD:6]: This was my second attempt at getting a working gRPC + Protobuf Rust compiler binary.

It works, but, ideally we would like to not write our own `main.rs`, but rather delegate it to an established Bazel Rust rules library - ie: `rules_rust` or other.

File: grpc/rust/BUILD [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    proto_plugin = "@vaticle_dependencies//library/crates:cargo_bin_protoc_gen_rust",
+)
+
+toolchain(
+    name = "proto_toolchain",
+    toolchain = ":proto_toolchain_impl",
+    toolchain_type = "@rules_rust//proto:toolchain_type",
+)
+
+rust_grpc_library(
```
Comment (alexjpwalker) [inline grpc/rust/BUILD:45]: This is our preferred way of creating a Rust gRPC library, and is symmetric with our Java, Python and NodeJS implementations. We delegate the work fully to `rules_rust`.

The only limitation is that it uses the `grpc` crate. Ideally, we would use the `tonic` crate, but I couldn't find a prebuilt Bazel rule that uses it.

**Review by alexjpwalker — COMMENTED** [review]
These changes look good.

**Review by lolski — APPROVED** [review]

**Review by haikalpribadi — APPROVED** [review]

**Pull request merged**